File: binarysearch.ts
type returns = {
  index: number;
  exist: boolean;
};

function binarySearch(nums: number[], target: number): returns {
  let l = 0;
  let hi = nums.length;
  do {
    const m = Math.floor(l + (hi - l) / 2);
    const v = nums[m];
    if (v === target) {
      return { index: v, exist: true };
    } else if (v > target) {
      hi = m;
    } else {
      l = m + 1;
    }
  } while (l < hi);

  return { index: -1, exist: false };
}

function bSearch(nums: number[], target: number): number {
  let l = 0;
  let r = nums.length - 1;

  do {
    const m = Math.floor(l + (r - l) / 2);
    const v = nums[m];

    if (v === target) {
      return v;
    } else if (v > target) {
      r = m;
    } else {
      l = m + 1;
    }
  } while (l < r);

  return -1;
}

File: binarysearch.go
package main

import "fmt"

func binarySearch(arr []int, target int) int {
    left, right := 0, len(arr)-1
    for left <= right {
        mid := (left + right) / 2
        if arr[mid] == target {
            return mid
        } else if arr[mid] < target {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }
    return -1
}

func main2() {
    nums := []int{1, 3, 5, 7, 9, 11}
    target := 7
    idx := binarySearch(nums, target)
    fmt.Println("Index:", idx)
}
